File: src/Binary/SearchInsertPosition.java
package Binary;

public class SearchInsertPosition {
    public int findMin(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;

        while (left < right) {
            int mid = (left + right) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

    public static void main(String[] args) {
        SearchInsertPosition s = new SearchInsertPosition();
        int nums[] = {0,1,2,3,4,5,6,7,8};
        System.out.println(s.findMin(nums, 4));
    }
}

File: py/SearchInsertPosition.py
# https://leetcode.com/problems/search-insert-position/description/
# Binary Search

from typing import List


class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums)  # target might be put at index n

        while left < right:
            mid = (right + left) // 2
            if nums[mid] < target:  # bisect left
                left = mid + 1
            else:
                right = mid

        return left


s = Solution()
nums = [1, 3, 5, 6]
target = 7
print(s.searchInsert(nums, target))
